### PR TITLE
first backtester implementation complete

### PR DESCRIPTION
**Pull request by cprovencher**

### CODE REVIEW

File: backtester/backtester.cpp [hunk 1]
```diff
@@ -151,7 +156,7 @@ void split(const string& line, string lineInfo[]) {
 	lineInfo[0] = pch; // time
 
 	pch = strtok_s(NULL, ",", &nextToken);
-	for (int i = 1; pch != NULL && i < 5; ++i) { // 5 = number of values we want in a line (time, open, high, low, close)
+	for (size_t i = 1; pch != NULL && i < 5; ++i) { // 5 = number of values we want in a line (time, open, high, low, close)
```
Comment (mrancourt) [inline backtester/backtester.cpp]: Could we extract the "5" as a const?

File: backtester/backtester.cpp [hunk 2]
```diff
@@ -197,27 +210,39 @@ void backtestAlgo(vector<Day>& days, json& config, ctpl::thread_pool& tp, const
 	}
 }
 
-void runParamCombos(int i, vector<string>& rangeNames, vector<vector<double>>& ranges, unordered_map<string, double>& params, const string& algoName, vector<Day>& days, ctpl::thread_pool& tp, vector<future<void>>& results) {
+void runParamCombos(size_t i, vector<string>& rangeNames, vector<vector<double>>& ranges, unordered_map<string, double>& params,
+					const string& algoName, vector<Day>& days, ctpl::thread_pool& tp, vector<future<void>>& results) {
 	double start = ranges[i][0];
 	double end = ranges[i][1];
 	double increment = ranges[i][2];
-	double margin = 0.3;
```
Comment (mrancourt) [inline backtester/backtester.cpp]: 0.3 as a const?

File: backtester/backtester.cpp [hunk 3]
```diff
@@ -226,15 +251,55 @@ void runParamCombos(int i, vector<string>& rangeNames, vector<vector<double>>& r
 	}
 }
 
+void simulateDay(double& currentCash, unordered_map<string, double>& params, size_t& currentTime, double& currentQuote, Action& action,
+	unique_ptr<Algorithm>& algo, Day& day, bool& activePositions, Result& result, Trade& trade, size_t& stockUnits) {
+	for (size_t i = 0; i < day.quotes.size(); ++i) {
+		if (currentCash < params["minimumCash"]) {
```
Comment (mrancourt) [inline backtester/backtester.cpp]: So if currentCash drops below minimumCash, we just stop processing anything for the whole day? I guess it's true that we don't want to buy more stocks, but we still might want to sell activePositions at a specific moment in the day provided by the algorithm right?

**Review by mrancourt — CHANGES_REQUESTED** [review]

File: backtester/backtester.cpp (hunk 3 above)
Comment (cprovencher) [reply backtester/backtester.cpp]: Whenever your cash drops below a threshold, it means the positions were sold in the last iteration and losses were incurred. Maybe we could add this code in the body of: if(action == sell && activePositions)

File: backtester/backtester.cpp (hunk 2 above)
Comment (cprovencher) [reply backtester/backtester.cpp]: This will be in the json now.

**Review by mrancourt — APPROVED** [review]

File: backtester/backtester.cpp (hunk 1 above)
Comment (cprovencher) [reply backtester/backtester.cpp]: Will set this as a global constant.

File: backtester/backtester.cpp (hunk 3 above)
Comment (mrancourt) [reply backtester/backtester.cpp]: Yeah breaking right after the last sell action that makes your cash drops below the threshold could be good. Or we can just leave it like that too...

File: backtester/algorithms/Simple.cpp [hunk 4]
```diff
@@ -1,14 +1,17 @@
 #include "Simple.h"
+#include <string>
 
-Simple::Simple(const unordered_map<string, double> params)
-{
+Simple::Simple() {
+	bought = false;
```
Comment (anadeau1) [inline backtester/algorithms/Simple.cpp]: Maybe we should initialize the parameters in a different way. It is possible to set them using this format:
Simle::Simple() 
         : bought(false)
{

}

That way you separate the logic from the variable initialization + they'll be easy to find

**Review by anadeau1 — APPROVED** [review]

**Pull request merged**